### ISSUE TITLE
Use of 'build' and 'image' config options together when using compose file prevents containers launching under local push

### ISSUE THREAD

**Issue by beaylott** [balena-io/balena-cli#1874]
In vanilla compose if 'image' and 'build' are used together the effect is to names/tags the built image with the provided image value (see: https://docs.docker.com/compose/compose-file/#build). 

When doing this with balena-cli (11.36.0) and balenaOS (2.47.0+rev1) pushing to a local dev device this seems to prevent the relevant containers from starting properly. 

e.g. something like this will not work:

```
services:
    service1:
        build:
            dockerfile: ./docker/Dockerfile
            context: ./service1
        image: myorg/service1:latest
```

but something like this will work:

```
services:
    service1:
        build:
            dockerfile: ./docker/Dockerfile
            context: ./service1
```

I would guess, in almost total ignorance, this might have something to do with the containers receiving different names from what is then expected by the supervisor?  But no error message is generated that I can see and the balena push logs just seem to indicate that installation is continuing.

Perhaps this is intended behaviour, if so would be good to document it somewhere.

**Comment by peteretep** [comment]
The same thing seems to happen when `container_name` is in the compose file.

**Comment by pdcastro** [comment]
Thank you for reporting this issue. 👍   I don't think it is intended behavior, more likely it is something that was not tested or anticipated during development (investigation required). I am labelling it a bug.

**Comment by joshbwlng** [comment]
Ran into the same problem when defining both `image` and `build` for services in `docker-compose.yml` and `balena push`ing to a local-mode device. balenaCLI debug logs:
```
...
[Logs]    [11/26/2020, 4:41:34 PM] Installing service 'tick local_image_tick:latest'
[Logs]    [11/26/2020, 4:41:34 PM] Installing service 'worker_1 local_image_worker_1:latest'
[Logs]    [11/26/2020, 4:41:34 PM] Installing service 'worker_2 local_image_worker_2:latest'
[Debug]   Device state not settled, retrying in 1000ms
[Debug]   Device state not settled, retrying in 1000ms
[Debug]   Device state not settled, retrying in 1000ms
[Debug]   Device state not settled, retrying in 1000ms
[Debug]   Device state not settled, retrying in 1000ms
[Debug]   Device state not settled, retrying in 1000ms
[Debug]   Device state not settled, retrying in 1000ms
[Debug]   Device state not settled, retrying in 1000ms
[Debug]   Device state not settled, retrying in 1000ms
[Logs]    [11/26/2020, 4:41:50 PM] Installing service 'tick local_image_tick:latest'
[Logs]    [11/26/2020, 4:41:50 PM] Installing service 'worker_1 local_image_worker_1:latest'
[Logs]    [11/26/2020, 4:41:50 PM] Installing service 'worker_2 local_image_worker_2:latest'
[Debug]   Device state not settled, retrying in 1000ms
[Debug]   Device state not settled, retrying in 1000ms
[Debug]   Device state not settled, retrying in 1000ms
[Debug]   Device state not settled, retrying in 1000ms
[Debug]   Device state not settled, retrying in 1000ms
[Debug]   Device state not settled, retrying in 1000ms
[Debug]   Device state not settled, retrying in 1000ms
[Debug]   Device state not settled, retrying in 1000ms
[Debug]   Device state not settled, retrying in 1000ms
[Debug]   Device state not settled, retrying in 1000ms
[Debug]   Device state not settled, retrying in 1000ms
[Debug]   Device state not settled, retrying in 1000ms
...
```

`balena.service` logs in the device:
```
...
Nov 26 07:42:59 localhost.localdomain c6badebdcb08[1141]: [event]   Event: Service install {"service":{"appId":1,"serviceId":8,"serviceName":"worker_2","releaseId":1}}
Nov 26 07:42:59 localhost.localdomain c6badebdcb08[1141]: [event]   Event: Service install {"service":{"appId":1,"serviceId":7,"serviceName":"worker_1","releaseId":1}}
Nov 26 07:42:59 localhost.localdomain c6badebdcb08[1141]: [error]   Scheduling another update attempt in 1000ms due to failure:  Error: Failed to apply state transition steps. (HTTP code 404) no such container - No such image: local_image_worker_2:latest  Steps:["start","start","start"]
Nov 26 07:42:59 localhost.localdomain c6badebdcb08[1141]: [error]         at /usr/src/app/dist/app.js:2:2958395
Nov 26 07:42:59 localhost.localdomain c6badebdcb08[1141]: [error]       at runNextTicks (internal/process/task_queues.js:62:5)
Nov 26 07:42:59 localhost.localdomain c6badebdcb08[1141]: [error]       at processImmediate (internal/timers.js:429:9)
Nov 26 07:42:59 localhost.localdomain c6badebdcb08[1141]: [event]   Event: Service install {"service":{"appId":1,"serviceId":5,"serviceName":"tick","releaseId":1}}
Nov 26 07:42:59 localhost.localdomain c6badebdcb08[1141]: [error]   Device state apply error Error: Failed to apply state transition steps. (HTTP code 404) no such container - No such image: local_image_worker_2:latest  Steps:["start","start","start"]
Nov 26 07:42:59 localhost.localdomain c6badebdcb08[1141]: [error]         at /usr/src/app/dist/app.js:2:2958395
Nov 26 07:42:59 localhost.localdomain c6badebdcb08[1141]: [error]       at runNextTicks (internal/process/task_queues.js:62:5)
Nov 26 07:42:59 localhost.localdomain c6badebdcb08[1141]: [error]       at processImmediate (internal/timers.js:429:9)
Nov 26 07:43:00 localhost.localdomain c6badebdcb08[1141]: [info]    Applying target state
Nov 26 07:43:00 localhost.localdomain c6badebdcb08[1141]: [event]   Event: Service install {"service":{"appId":1,"serviceId":5,"serviceName":"tick","releaseId":1}}
Nov 26 07:43:00 localhost.localdomain c6badebdcb08[1141]: [event]   Event: Service install {"service":{"appId":1,"serviceId":7,"serviceName":"worker_1","releaseId":1}}
Nov 26 07:43:00 localhost.localdomain c6badebdcb08[1141]: [event]   Event: Service install {"service":{"appId":1,"serviceId":8,"serviceName":"worker_2","releaseId":1}}
Nov 26 07:43:00 localhost.localdomain c6badebdcb08[1141]: [error]   Updating failed, but there's another update scheduled immediately:  Error: Failed to apply state transition steps. (HTTP code 404) no such container - No such image: local_image_tick:latest  Steps:["start","start","start"]
Nov 26 07:43:00 localhost.localdomain c6badebdcb08[1141]: [error]         at /usr/src/app/dist/app.js:2:2958395
Nov 26 07:43:00 localhost.localdomain c6badebdcb08[1141]: [error]   Device state apply error Error: Failed to apply state transition steps. (HTTP code 404) no such container - No such image: local_image_tick:latest  Steps:["start","start","start"]
Nov 26 07:43:00 localhost.localdomain c6badebdcb08[1141]: [error]         at /usr/src/app/dist/app.js:2:2958395
...
```

Internal conversation: https://www.flowdock.com/app/rulemotion/r-beginners/threads/FbJ9LRMrYFmsoo5AEEqiWBzYcLD

**Comment by LucianBuzzo** [comment]
@peteretep I recently dug into the original issue posted here and was able to find a solution, but I was unable to reproduce the issue you describe with `container_name`. If possible would you be able to open a separate issue with details on how to reproduce the bug, ideally with an example `docker-compose.yml` or even a GitHub repo I could test with?

**Comment by peteretep** [comment]
Sorry, I'm no longer working as a software engineer so won't be able to
look at this.

**Comment by pdcastro** [comment]
I understand this issue was resolved in CLI v12.54.2 (by PR #2395), hence closing it. If anyone finds that's not the case, please add a comment here and we'll reopen the issue and investigate it further.
